src/crypto/NoteEncryption.cpp

```diff
@@ -0,0 +1,161 @@
+#include "NoteEncryption.hpp"
+#include "crypto/sha512.h"
+#include <stdexcept>
+#include "sodium.h"
+#include <boost/static_assert.hpp>
+#include "prf.h"
+#include "zerocash/Zerocash.h"
+
+#define NOTEENCRYPTION_CIPHER_KEYSIZE 32
+
+void clamp_curve25519(unsigned char *key)
+{
+    key[0] &= 248;
+    key[31] &= 127;
+    key[31] |= 64;
+}
+
+void KDF(unsigned char K[NOTEENCRYPTION_CIPHER_KEYSIZE],
+    const uint256 &dhsecret,
+    const uint256 &epk,
+    const uint256 &pk_enc,
+    const uint256 &seed,
+    unsigned char nonce
+   )
+{
+    unsigned char K_tmp[64];
+    CSHA512 hasher;
+
+    // Trim the first byte from seed
+    std::vector<unsigned char> first_248(seed.begin() + 1, seed.end());
+
+    if (nonce >= 2) {
+        throw std::logic_error("KDF nonce space exceeded");
+    }
+
+    // First bit is the 'nonce' bit or `i`
+    // Second bit is zero
+    first_248[0] &= 0xFC;
+    first_248[0] |= nonce;
+
+    hasher.Write(&first_248[0], 31);
+    hasher.Write(dhsecret.begin(), crypto_scalarmult_BYTES);
+    hasher.Write(epk.begin(), crypto_scalarmult_BYTES);
+    hasher.Write(pk_enc.begin(), crypto_scalarmult_BYTES);
+    hasher.Finalize(K_tmp);
+
+    memcpy(K, K_tmp + 32, 32);
+}
+
+template<size_t MLEN>
+NoteEncryption<MLEN>::NoteEncryption(uint256 seed) : nonce(0), seed(seed) {
+    // All of this code assumes crypto_scalarmult_BYTES is 32
+    // There's no reason that will _ever_ change, but for
+    // completeness purposes, let's check anyway.
+    BOOST_STATIC_ASSERT(32 == crypto_scalarmult_BYTES);
+    BOOST_STATIC_ASSERT(32 == crypto_scalarmult_SCALARBYTES);
+    BOOST_STATIC_ASSERT(NOTEENCRYPTION_AUTH_BYTES == crypto_aead_chacha20poly1305_ABYTES);
+
+    // Create the ephemeral keypair
+    randombytes_buf(esk.begin(), crypto_scalarmult_SCALARBYTES);
+    epk = generate_pubkey(esk);
+}
+
+template<size_t MLEN>
+NoteEncryption<MLEN>::~NoteEncryption() {
+    sodium_memzero(epk.begin(), crypto_scalarmult_BYTES);
+    sodium_memzero(esk.begin(), crypto_scalarmult_SCALARBYTES);
+}
+
+template<size_t MLEN>
+typename NoteEncryption<MLEN>::Ciphertext NoteEncryption<MLEN>::encrypt
+                                          (const uint256 &pk_enc,
+                                           const NoteEncryption<MLEN>::Plaintext &message
+                                          )
+{
+    uint256 dhsecret;
```

Comment:
I actually don't see much point in zeroing epk/esk anyway, I guess it can't hurt, but there are plenty of other places we don't zero things. It pretty much means nothing and I don't remember why I did it -- probably because if NoteEncryption gets heap allocated for some reason, I would want the data securely destroyed, something I don't usually care about for stack allocated stuff.

@daira's input on this would be helpful.